metrics/fidelity/utility.py

```diff
@@ -0,0 +1,174 @@
+    if dataset_level:
+        mean_difference = np.mean(mean_difference)
+    if fidelity_type == "sufficiency":
+        return 1 - mean_difference
+    else:
+        return mean_difference
+
+
+def itemwize_normalization(fidelity, null_difference, fidelity_type, clip, eps):
+    """
+    Normalize a fidelity value given the null difference value
+    :param fidelity: sufficiency or comprehensiveness metric
+    :param null_difference: max(0, P(y_hat/x) - P(y_hat/x, 0)); alpha = 0
+    :param fidelity_type: sufficiency or comprehensiveness
+    :param clip: true for clipping the fidelity values between 0 and 1
+    :param eps: denominator threshold when it tends to 0
+    :return: normalized fidelity
+    """
+    try:
+        if fidelity_type == "sufficiency":
+            numerator = fidelity - (1 - null_difference)
+            denominator = 1 - (1 - null_difference)
+        else:
+            numerator = fidelity
+            denominator = null_difference
+
+        if np.abs(denominator) < eps:
+            result = 0
+        else:
+            result = numerator / denominator
+    except Exception as e:
+        logger.info(e)
+        return 0
+
+    if clip:
+        return np.clip(result, 0, 1)
+    else:
+        return result
+
+
+def normalization(fidelity, null_difference, fidelity_type="sufficiency", clip=True, eps=1e-5):
+    """
+    Normalize a list of fidelity values given the null difference values
+    :param fidelity: sufficiency or comprehensiveness metric
+    :param null_difference: max(0, P(y_hat/x) - P(y_hat/x, 0)); alpha = 0
+    :param fidelity_type: sufficiency or comprehensiveness
+    :param clip: true for clipping the fidelity values between 0 and 1
+    :param eps: denominator threshold when it tends to 0
+    :return: normalized fidelity
+    """
+    normalized_fidelity = []
+    for idx, itemwize_fidelity in enumerate(fidelity):
+        normalized_fidelity.append(
+            itemwize_normalization(
+                fidelity=itemwize_fidelity,
+                null_difference=null_difference[idx],
+                fidelity_type=fidelity_type,
+                clip=clip,
+                eps=eps,
+            )
+        )
+
+    return normalized_fidelity
+
+
+def compute_predictions(input_ids, model, attention_masks=None):
+    """
+    Compute the prediction for given input_ids, model and attention masks
+    This function returns numpy arrays of labels and probabilities associated with that label
+    :param input_ids: tensor of input ids
+    :param model: pytorch nn Module
+    :param attention_masks: tensor of attention masks
+    :return: numpy array of labels and probabilities associated with that label
+    """
+    if attention_masks is not None:
+        result = model.forward(input_ids=input_ids.to(model.device), attention_mask=attention_masks.to(model.device))
+    else:
+        result = model.forward(input_ids=input_ids.to(model.device))
+
+    if model.device == "cuda":
```

Comment:
So, OUR models have a forward function that returns a dictionary with a "py_index" key and a "probs" key. But you can't assume that any arbitrary model will have these things. So you need to make this function smarter about how it derives the predicted class and predicted class probabilities from the model output. 

If you look at the documentation for BertForSequenceClassification (https://huggingface.co/transformers/model_doc/bert.html#transformers.BertForSequenceClassification), you can see that their forward function returns the logits as the second element in a tuple. So the logic in your function here probably ought to be something like, if the return value is a tuple, assume the second element is logits and process them from there, while if it is a dictionary, look first for a "py_index" and a "probs", and if they are not present then look for a "logits", and if can't find that then throw an informative exception.

Also, give the calling user an argument to specify what key (integer or string) is the logits (logit_key) or probs (prob_key) in the model output, so they can use their own model with this code. Pass these arguments through to this function from the top level metric code.